File: Backtracking/676. Implement Magic Dictionary.py

"""
676. Implement Magic Dictionary
"""

import collections
class MagicDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
    def _genneighbours(self):
        for word in self.dict:
            for i in range(len(word)):
                yield word[:i]+"*"+word[i+1:]

    def buildDict(self, dict):
        """
        Build a dictionary through a list of words
        :type dict: List[str]
        :rtype: void
        """
        self.dict = set(dict)
        self.map = collections.Counter(x for x in self._genneighbours())
        

    def search(self, word):
        """
        Returns if there is any word in the trie that equals to the given word after modifying exactly one character
        :type word: str
        :rtype: bool
        """
        return any(self.map[word[:i]+"*"+word[i+1:]] > 1 or 
                  self.map[word[:i]+"*"+word[i+1:]] == 1 and word not in self.dict 
                  for i in range(len(word)) )

        


# Your MagicDictionary object will be instantiated and called as such:
# obj = MagicDictionary()
# obj.buildDict(dict)
# param_2 = obj.search(word)
        


# Your MagicDictionary object will be instantiated and called as such:
# obj = MagicDictionary()
# obj.buildDict(dict)
# param_2 = obj.search(word)



#2020
class MagicDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = {}

        
    def buildDict(self, dict: List[str]) -> None:
        """
        Build a dictionary through a list of words
        """
        
        for word in dict:  
            node = self.root 
            for i in word: 
                node = node.setdefault(i, {})
            node[None] = True 

    def search(self, word: str) -> bool:
        """
        Returns if there is any word in the trie that equals to the given word after modifying exactly one character
        """
        def dfs(node, s, oneDiff): #always has child for node, 不用check if node is empty, 因为for loop中阻止了没有child的可能性
            if not s:
                   return oneDiff and None in node #if already oneDiff, current node is leaf, return true 
            
            if s[0] not in node and oneDiff:# current position is different, and before has one differnt
                return False
            else: 
                if s[0] in node:
                   if dfs(node[s[0]],s[1:], oneDiff): 
                        return True 
                   elif oneDiff: #above return False, and already have one difference, cannot modify current position, return False 
                        return False 
                for k,nxt in node.items():
                    if nxt != True and k != s[0] and dfs(nxt, s[1:], 1): #nxt == True means current leaf and don't have child
                        return True 
                return False 
        return dfs(self.root, word, 0)

File: Backtracking/676. Implement Magic Dictionary.cpp

/*
676. Implement Magic Dictionary

Implement a magic directory with buildDict, and search methods.

For the method buildDict, you'll be given a list of non-repetitive words to build a dictionary.

For the method search, you'll be given a word, and judge whether if you modify exactly one character into another character in this word,
 the modified word is in the dictionary you just built.

Example 1:
Input: buildDict(["hello", "leetcode"]), Output: Null
Input: search("hello"), Output: False
Input: search("hhllo"), Output: True
Input: search("hell"), Output: False
Input: search("leetcoded"), Output: False
Note:
You may assume that all the inputs are consist of lowercase letters a-z.
For contest purpose, the test data is rather small by now. You could think about highly efficient algorithm after the contest.
Please remember to RESET your class variables declared in class MagicDictionary, as static/class variables are persisted across 
multiple test cases. Please see here for more details.

*/


class MagicDictionary {
public:
	/** Initialize your data structure here. */
	MagicDictionary() {

	}

	/** Build a dictionary through a list of words */
	void buildDict(vector<string> dict) {
		map.insert(dict.begin(), dict.end());
	}

	/** Returns if there is any word in the trie that equals to the given word after modifying exactly one character */
	bool search(string word) {
		for (int i = 0; i < word.size(); i++) {
			for (int j = 0; j <= 25; j++){
				if (word[i] == 'a' + j) continue;
				char k = word[i];
				word[i] = 'a' + j;
				if (map.count(word) > 0) return true;
				word[i] = k;
			}
		}
		return false;
	}

	unordered_set<string>map;
};



/**
 * 
 * instead search every chars from A-Z, search hello as *ello, h*llo
 * /

class MagicDictionary {
public:
	/** Initialize your data structure here. */
	MagicDictionary() {

	}

	/** Build a dictionary through a list of words */
	void buildDict(vector<string> dict) {
		map.insert(dict.begin(), dict.end());
        for(auto word: dict){
            int len = word.size();
            for(int i = 0; i<len; i++){
                char ch = word[i];
                word[i] = '*';
                transform[len][word]++;
                word[i] = ch;
            }
        }
	}

	/** Returns if there is any word in the trie that equals to the given word after modifying exactly one character */
	bool search(string word) {
		int len = word.size();
        if(transform.count(len)==0 ) return false;
        for(int i = 0; i<len; i++){
            string tword = word;
            tword[i] = '*';
            if(transform[len].find(tword)!=transform[len].end() && (transform[len][tword]>1 || transform[len][tword]==1 && map.count(word)==0)) return true;
        }
		return false;
	}

	unordered_set<string>map;
    unordered_map<int,unordered_map<string,int>>transform;
};


/**
 * Your MagicDictionary object will be instantiated and called as such:
 * MagicDictionary obj = new MagicDictionary();
 * obj.buildDict(dict);
 * bool param_2 = obj.search(word);
 */